### PR TITLE
Split tokens detection from balances query

### PR DESCRIPTION
**Pull request by nebolax**

### CODE REVIEW

**Review by LefterisJP — COMMENTED** [review]
hey @nebolax thanks for making a smaller PR to ask.

So the idea would be to split those 2 in different endpoints.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 1]
```diff
@@ -61,65 +59,82 @@
...
+        chunks: Iterable[List[EthereumToken]],
+        querying_etherscan: bool,
+    ) -> Dict[EthereumToken, FVal]:
+        total_token_amounts: Dict[EthereumToken, FVal] = defaultdict(FVal)
+        for chunk in chunks:
+            if querying_etherscan is True:
+                call_order = [ETHERSCAN_NODE]
+            else:
+                call_order = self.ethereum.default_call_order()
+            new_token_amounts = self._get_token_amounts(
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:116]: adjust amount to balances here and below

File: rotkehlchen/chain/ethereum/tokens.py [hunk 2]
```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
...
-            account=address,
-            call_order=call_order,
-        )
-        for token_identifier, value in ret.items():
-            token = EthereumToken.from_identifier(token_identifier)
-            if token is None:  # should not happen
-                log.warning(
-                    f'Could not initialize token with identifier {token_identifier}. '
-                    f'Should not happen. Skipping its token balance query',
+        for account in accounts:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:167]: Change it up a bit here.


Since only `n` and `querying_etherscan` is what is differentiated here simply set the variables appropriately depending on `self.ethereum.connected_to_any_web3()` and then after the conditional call `self._query_chunks` with the customized variables.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 3]
```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
...
-                )
-                balances[token.identifier] = normalized_amount
-        return balances
+        token_usd_price: Dict[EthereumToken, Price] = {}
+        result = {}
+
+        for address in addresses:
+            with self.db.conn.read_ctx() as cursor:
+                saved_list = self.db.get_tokens_for_address(cursor, address=address)
+            if saved_list is None or len(saved_list) == 0:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:201]: Can't you make sure that the list is either `None` if empty or is guaranteed to at least have 1 token if returned from the DB? Since we are the ones who create it, this should be possible and this way we can save an if check per address.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 4]
```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
...
-            if token_amount != 0:
-                normalized_amount = token_normalized_value(token_amount, token)
-                log.debug(
-                    f'Found {token.symbol}({token.ethereum_address}) token balance for '
-                    f'{account} and amount {normalized_amount}',
-                )
-                balances[token.identifier] = normalized_amount
-        return balances
+        token_usd_price: Dict[EthereumToken, Price] = {}
+        result = {}
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:196]: rename for code clarity and readability

```suggestion
        address_to_balance = {}
```

File: rotkehlchen/chain/ethereum/tokens.py [hunk 5]
```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
...
+
+            token_amounts = self._get_token_amounts(
+                account=address,
+                tokens=saved_list,
+                call_order=None,  # use defaults
+            )
+            result[address] = token_amounts
+
+            for token in token_amounts.keys() - token_usd_price:
+                token_usd_price[token] = Inquirer.find_usd_price(asset=token)
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:212]: This is an example of badly written code. There is too much happening and it's all implicit and requires the reader to know too many things. Good code is simple and readable.

1. You need to know that `keys()` returns a Dictionary view object that also happens to be set-like
2. You need to know that the `token_usd_price` gets coerced into being a set.
3. So in the end you have a set subtraction here.

The way to make it more readable is either assign `token_amounts.keys() - token_usd_price` into a variable named `tokens_without_price` and then iterate that variable or explain what you are doing above the loop with a comment.

`# Get the subset of tokens we don't yet know the price for`

------

But most importantly this is probably not needed.

`Inquirer.find_usd_price()` has a timed cache for the current price. Check its implementation. So the optimization you are doing is not required.

Also you should not query prices inside the loop. Instead gather all tokens at some point and query prices at the very end.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 6]
```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
...
+        token_usd_price: Dict[EthereumToken, Price] = {}
+        result = {}
+
+        for address in addresses:
+            with self.db.conn.read_ctx() as cursor:
+                saved_list = self.db.get_tokens_for_address(cursor, address=address)
+            if saved_list is None or len(saved_list) == 0:
+                continue  # Do not query if we know the address has no tokens
+
+            token_amounts = self._get_token_amounts(
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:204]: Perhaps once the PR is ready, another optimization to do is the following:

Instead of doing the eth.call() for address -> tokens, combine it if there is enough space for combinations of address -> tokens, address -> tokens.

Why? If you have 20 addresses all with 1 token, you still end up doing 20 blockchain queries. While you can definitely still do only 1.

This is also an improvement the old code was lacking -- and we wanted to do for a long time, so why not do it now?

File: rotkehlchen/api/rest.py [hunk 7]
```diff
@@ -4416,3 +4417,41 @@ def search_for_names_everywhere(self, addresses: List[ChecksumEthAddress]) -> Re
...
+            addresses = self.rotkehlchen.chain_manager.accounts.eth
+        if async_query is True:
+            return self._query_async(
+                command=self._detect_ethereum_tokens,
+                addresses=addresses,
+            )
+
+        response = self._detect_ethereum_tokens(
+            addresses=addresses,
+        )
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:4454]: If code fits in one line put it there
```suggestion
        response = self._detect_ethereum_tokens(addresses=addresses)
```

File: rotkehlchen/api/v1/schemas.py [hunk 8]
```diff
@@ -2318,3 +2318,7 @@ def validate_schema(  # pylint: disable=no-self-use
                 message='Can\'t delete the etherscan node',
                 field_name='identifier',
             )
+
+
+class DetectTokensSchema(AsyncQueryArgumentSchema):
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:2323]: This is the exact same schema as `AddressbookAddressesSchema` uses. Abstract it into a common schema name: `OptionalAddressesListSchema` and then use it both in `AddressbookAddressesSchema` and in the one you are making here. Using it would probably simply mean inheriting, essentially like implementing an interface.

File: rotkehlchen/db/dbhandler.py [hunk 9]
```diff
@@ -1320,21 +1320,18 @@ def remove_blockchain_accounts(
...
-    def _get_address_details_if_time(
+    def _get_address_details(
             self,
             cursor: 'DBCursor',
             address: ChecksumEthAddress,
-            current_time: Timestamp,
     ) -> Optional[Dict[str, Any]]:
         query = cursor.execute(
-            'SELECT tokens_list, time FROM ethereum_accounts_details WHERE account = ?',
+            'SELECT tokens_list FROM ethereum_accounts_details WHERE account = ?',
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:1329]: If you code it this way then the `time` argument is no longer needed. I think it should be. And we should even show to the user for each address when is the last time tokens were asked to be detected.

Remember this is only for the "brute-force" detection where we essentially iterate every token we know. This is always gonna be slow.

The other nice type of detection we will do is process all transaction events and there we can see all tokens to touch an address. We can even (in theory) calculate balances this way.

**Review by LefterisJP — COMMENTED** [review]
Left you lots of comments.

This PR basically only splits out the detection in a separate api endpoint.

It's okay for a start but we would need to have this also implemented in the frontend if this is to be even remotely viable in develop. And for that you need to talk with the frontend about how this should look like. What kind of data would they need?

Frontend should:
- Show what tokens are detected.
- Show when was the last time an address had its token detection run.

We have this data. You need to figure out how/where to return it.

Then we should also probably think on how to do the transactions processing to extract (among other things) tokens touched by each address so that token detection also happens in a bit more error prone way, but automatically.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 10]
```diff
@@ -61,65 +59,82 @@
 
 
 class EthTokens():
-
     def __init__(self, database: DBHandler, ethereum: EthereumManager):
         self.db = database
         self.ethereum = ethereum
 
-    def detect_tokens_for_address(
+    def _get_token_amounts(
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:66]: Done

File: rotkehlchen/chain/ethereum/tokens.py [hunk 11]
```diff
@@ -61,65 +59,82 @@
...
-    def query_tokens_for_addresses(
-            self,
-            addresses: List[ChecksumEthAddress],
-            force_detection: bool,
-    ) -> TokensReturn:
-        """Queries/detects token balances for a list of addresses
-
-        If an address's tokens were recently autodetected they are not detected again but the
-        balances are simply queried. Unless force_detection is True.
+        """Queries amounts of multiple tokens for an account
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:72]: Got it

File: rotkehlchen/chain/ethereum/tokens.py (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:116]: Done

File: rotkehlchen/chain/ethereum/tokens.py [hunk 12]
```diff
@@ -61,65 +59,82 @@
...
+            total_token_amounts = combine_dicts(total_token_amounts, new_token_amounts)
+        return total_token_amounts
+
+    def detect_tokens(
+        self,
+        write_cursor: 'DBCursor',
+        accounts: List[ChecksumEthAddress],
+    ) -> None:
+        """
+        Detect tokens for the specified account.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:130]: Done

File: rotkehlchen/chain/ethereum/tokens.py (hunk 2 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:167]: Done

File: rotkehlchen/chain/ethereum/tokens.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:196]: Done

File: rotkehlchen/chain/ethereum/tokens.py (hunk 3 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:201]: Moved this check to `get_tokens_for_address`

File: rotkehlchen/chain/ethereum/tokens.py (hunk 5 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:212]: Got it, done, thanks for the explanation

File: rotkehlchen/chain/manager.py [hunk 13]
```diff
@@ -604,13 +604,11 @@ def check_accounts_exist(
     def query_balances(
             self,
             blockchain: Optional[SupportedBlockchain] = None,
-            force_token_detection: bool = False,
+            beaconchain_fetch_eth1: bool = False,
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/manager.py:607]: Added

File: rotkehlchen/api/rest.py (hunk 7 above)
Comment (nebolax) [reply rotkehlchen/api/rest.py:4454]: Done

File: rotkehlchen/api/v1/schemas.py (hunk 8 above)
Comment (nebolax) [reply rotkehlchen/api/v1/schemas.py:2323]: Done

File: rotkehlchen/chain/ethereum/tokens.py (hunk 6 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:204]: Yeah, I see. Will a new be better for this?

File: rotkehlchen/tests/integration/test_blockchain.py [hunk 14]
```diff
@@ -104,11 +106,19 @@ def mock_add_defi_balances_to_token_and_totals():
...
+                only_cache=False,
+                accounts=[addr1, addr2],
+            )
+
     assert addr1 in blockchain.accounts.eth
 
     with etherscan_patch, ethtokens_max_chunks_patch, defi_balances_mock, add_defi_mock, beaconchain_patch:  # noqa: E501
         greenlets = [
-            gevent.spawn_later(0.01 * x, blockchain.query_ethereum_balances, False)
+            gevent.spawn_later(0.01 * x, blockchain.query_ethereum_balances, ignore_cache=True)
```
Comment (nebolax) [inline rotkehlchen/tests/integration/test_blockchain.py:121]: This test was useless before because cache was used instead of calling `query_ethereum_balances`. Spent quite some time figuring it out.

File: rotkehlchen/api/rest.py [hunk 15]
```diff
@@ -4416,3 +4417,49 @@ def search_for_names_everywhere(self, addresses: List[ChecksumEthAddress]) -> Re
...
+                )
+        except (RemoteError, BadFunctionCallOutput) as e:
+            return wrap_in_fail_result(message=str(e), status_code=HTTPStatus.CONFLICT)
+        result = {}
+        for account, (tokens, last_update_ts) in account_tokens_info.items():
+            result[account] = {
+                'tokens': [token.identifier for token in tokens] if tokens is not None else None,
+                'last_update_timestamp': last_update_ts,
+            }
+        return _wrap_in_ok_result(result)
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:4445]: _wrap_in_ok_result does not put a status code! You need a status code here, so either do the wrapping manually or add a default argument to `_wrap_in_ok_result` just like `wrap_in_fail_result` has

File: rotkehlchen/tests/integration/test_blockchain.py (hunk 14 above)
Comment (LefterisJP) [reply rotkehlchen/tests/integration/test_blockchain.py:121]: It was not useless. This test was failing before. There was a bug where cache was not only ignored but something triggered an exception in the code.

With your change what are you achieving? You are just making it take a lot more time since you do 5 blockchain queries without any cache.

File: rotkehlchen/api/v1/resources.py [hunk 16]
```diff
@@ -2568,7 +2569,7 @@ def get(self) -> Response:
 
 
 class AddressbookResource(BaseMethodView):
-    post_delete_schema = AddressbookAddressesSchema()
+    post_delete_schema = OptionalAddressesListSchema()
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:2572]: Nope, this will break (and by checking the tests it has) all address book tests.

File: rotkehlchen/api/v1/schemas.py [hunk 17]
```diff
@@ -2147,10 +2147,6 @@ class BaseAddressbookSchema(Schema):
     book_type = SerializableEnumField(enum_class=AddressbookType, required=True)
 
 
-class AddressbookAddressesSchema(BaseAddressbookSchema):
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:2150]: This is a mistake. This was previous extending `BaseAddressbookSchema` and you are replacing this directly with `OptionalAddressesListSchema` which is a mistake.

You are removing the `book_type` argument and adding an unneeded `async_query` argument.

What you need to do is keep this but have an `OptionalAddressesListSchema` **without** the `AsyncQueryArgumentSchema`. And extend this here too.

And later down in `DetectTokensSchema` also add `AsyncQueryArgumentSchema`.

File: rotkehlchen/api/v1/schemas.py [hunk 18]
```diff
@@ -2318,3 +2314,14 @@ def validate_schema(  # pylint: disable=no-self-use
                 message='Can\'t delete the etherscan node',
                 field_name='identifier',
             )
+
+
+class OptionalAddressesListSchema(AsyncQueryArgumentSchema):
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:2319]: Do this as I explained above
```suggestion
class OptionalAddressesListSchema(Schema):
```

File: rotkehlchen/api/v1/schemas.py [hunk 19]
```diff
@@ -2318,3 +2314,14 @@ def validate_schema(  # pylint: disable=no-self-use
...
+    addresses = fields.List(EthereumAddressField(required=True), load_default=None)
+
+
+class DetectTokensSchema(
+    OptionalAddressesListSchema,
+    OnlyCacheQuerySchema,
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:2325]: As I explained above.

```suggestion
    OnlyCacheQuerySchema,
    AsyncQueryArgumentSchema,
```

File: rotkehlchen/chain/ethereum/tokens.py [hunk 20]
```diff
@@ -61,64 +64,92 @@
...
-    ) -> TokensReturn:
-        """Queries/detects token balances for a list of addresses
-
-        If an address's tokens were recently autodetected they are not detected again but the
-        balances are simply queried. Unless force_detection is True.
+    def _query_chunks(
+        self,
+        account: ChecksumEthAddress,
+        chunks: Iterable[List[EthereumToken]],
+        querying_etherscan: bool,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:113]: Please configure your editor to follow the indentation style. + 1 indentation on these arguments.

```suggestion
        self,
        account: ChecksumEthAddress,
        chunks: Iterable[List[EthereumToken]],
        querying_etherscan: bool,
```

File: rotkehlchen/chain/ethereum/tokens.py [hunk 21]
```diff
@@ -61,64 +64,92 @@
...
+                call_order=call_order,
+            )
+            total_token_balances = combine_dicts(total_token_balances, new_token_balances)
+        return total_token_balances
+
+    def detect_tokens(
+        self,
+        write_cursor: 'DBCursor',
+        only_cache: bool,
+        accounts: List[ChecksumEthAddress],
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:133]: Please configure your editor to follow the indentation style. + 1 indentation on these arguments.


```suggestion
        self,
        write_cursor: 'DBCursor',
        only_cache: bool,
        accounts: List[ChecksumEthAddress],
```

File: rotkehlchen/chain/ethereum/tokens.py [hunk 22]
```diff
@@ -61,64 +64,92 @@
...
+    ) -> DetectedTokensType:
+        """
+        Detect tokens for the given accounts.
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:136]: Explain in the docstring what the `only_cache` argument does and what the function is supposed to return.

Make it a habit to write docstrings.

File: rotkehlchen/chain/ethereum/tokens.py [hunk 23]
```diff
@@ -61,64 +64,92 @@
...
+        - BadFunctionCallOutput if a local node is used and the contract for the
+          token has no code. That means the chain is not synced
         """
-        log.debug(
-            'Querying/detecting token balances for all addresses',
-            force_detection=force_detection,
-        )
-        with self.db.user_write() as cursor:
-            ignored_assets = self.db.get_ignored_assets(cursor)
+        if only_cache is True:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:144]: Also wouldn't the logic here be much better if it:

1. If not cache, query the chain and write to the DB
2. And then in the end always read and return the result from the DB.

This is what we try to do in the places where DB cache is involve.d

File: rotkehlchen/chain/ethereum/tokens.py [hunk 24]
```diff
@@ -141,116 +172,65 @@ def query_tokens_for_addresses(
...
-            call_order: Optional[Sequence[WeightedNode]],
-    ) -> Dict[str, FVal]:
-        """Queries balances of multiple tokens for an account
-
-        Return a dictionary with keys being tokens and value a dictionary of
-        account to balances
+            addresses: List[ChecksumEthAddress],
+    ) -> TokenBalancesType:
+        """Queries token balances for a list of addresses
+        Returns the token balances of each address and the usd prices of the tokens
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:207]:
```suggestion
        Returns the token balances of each address and the usd prices of the tokens.
```

File: rotkehlchen/chain/ethereum/tokens.py [hunk 25]
```diff
@@ -141,116 +172,65 @@ def query_tokens_for_addresses(
...
-                    f'Found {token.symbol}({token.ethereum_address}) token balance for '
-                    f'{account} and amount {normalized_amount}',
-                )
-                balances[token.identifier] = normalized_amount
-        return balances
+        address_to_balance: Dict[ChecksumEthAddress, Dict[EthereumToken, FVal]] = {}
+        all_tokens = set()
+
+        for address in addresses:
+            with self.db.conn.read_ctx() as cursor:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/tokens.py:219]: Move the db cursor context manager out of the loop. Why would you recreate a cursor and a context manager for each address and loop iteration?

File: rotkehlchen/db/dbhandler.py [hunk 26]
```diff
@@ -1344,38 +1342,40 @@ def _get_address_details_if_time(
...
 
         if not isinstance(tokens_list, list):
             # This should never happen
             self.msg_aggregator.add_warning(
                 f'Found non-list tokens_list {json_ret} in the DB for {address}.',
             )
-            return None
+            return None, last_update_ts
+
+        if len(tokens_list) == 0:
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:1377]: This is not what I asked you to do though You just moved the check but same amount of checks will happen each time. https://github.com/rotki/rotki/pull/4536#discussion_r928169962

But I think for now it's fine. I was thinking of making sure to write null in case of empty list, but looking at the table and how it's used I see we have a dictionary entry there that is never null, with multiple keys, so don't wanna touch it now.

File: rotkehlchen/tests/integration/test_blockchain.py (hunk 14 above)
Comment (LefterisJP) [reply rotkehlchen/tests/integration/test_blockchain.py:121]: It was not useless. This test was failing before, with the way the codebase was when this was written. There was a bug where cache was not only ignored but something triggered an exception in the code.

With your change what are you achieving? You are just making it take a lot more time since you do 5 blockchain queries without any cache.

File: rotkehlchen/db/dbhandler.py (hunk 26 above)
Comment (nebolax) [reply rotkehlchen/db/dbhandler.py:1377]: Yeah, I basically didn't find a way to get rid of this check

File: rotkehlchen/chain/ethereum/tokens.py (hunk 25 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:219]: Makes sense

File: rotkehlchen/chain/ethereum/tokens.py (hunk 23 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:144]: Sounds good

File: rotkehlchen/chain/ethereum/tokens.py (hunk 22 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:136]: Ok

File: rotkehlchen/chain/ethereum/tokens.py (hunk 21 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:133]: Ok

File: rotkehlchen/chain/ethereum/tokens.py (hunk 20 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py:113]: Ok

File: rotkehlchen/api/v1/schemas.py (hunk 17 above)
Comment (nebolax) [reply rotkehlchen/api/v1/schemas.py:2150]: Oh yeah yeah, I understand it. Just was a bit inattentive in the evening after spending my brain resource onto understanding why the test (with cache) was failing.

File: rotkehlchen/api/v1/resources.py (hunk 16 above)
Comment (nebolax) [reply rotkehlchen/api/v1/resources.py:2572]: Yeah, yeah

File: rotkehlchen/api/rest.py [hunk 27]
```diff
@@ -4416,3 +4417,49 @@ def search_for_names_everywhere(self, addresses: List[ChecksumEthAddress]) -> Re
...
+        try:
+            with self.rotkehlchen.data.db.user_write() as cursor:
+                account_tokens_info = ethtokens.detect_tokens(
+                    write_cursor=cursor,
+                    only_cache=only_cache,
+                    accounts=addresses,
+                )
+        except (RemoteError, BadFunctionCallOutput) as e:
+            return wrap_in_fail_result(message=str(e), status_code=HTTPStatus.CONFLICT)
+        result = {}
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/api/rest.py:4439]: Ok

File: rotkehlchen/api/rest.py (hunk 15 above)
Comment (nebolax) [reply rotkehlchen/api/rest.py:4445]: `_get_status_code_from_async_response` which is used further down the line has the `status_code` parameter with default value `OK`.

But no problem, I will add the status code here too to be more explicit.

File: docs/api.rst [hunk 28]
```diff
@@ -2516,7 +2516,7 @@ Querying onchain balances
       Host: localhost:5042
 
    :reqjson bool async_query: Boolean denoting whether this is an asynchronous query or not
-   :reqjson bool ignore_cache: Boolean denoting whether to ignore the cache for this query or not.
+   :reqjson bool ignore_cache: Boolean denoting whether to ignore balance amounts cache for this query or not.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/api.rst:2519]: Hm... yes, probably it is not needed

File: rotkehlchen/db/dbhandler.py (hunk 9 above)
Comment (nebolax) [reply rotkehlchen/db/dbhandler.py]: Yeah, added time to the response

File: docs/api.rst [hunk 29]
```diff
@@ -2823,6 +2823,44 @@ Querying owned assets
...
+
+   **Example Response**:
+
+   .. sourcecode:: http
+
+      HTTP/1.1 200 OK
+      Content-Type: application/json
+
+      {
+          "result": true,
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/api.rst]: Done

File: rotkehlchen/chain/ethereum/tokens.py (hunk 6 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/tokens.py]: Will leave this convo open to don't forget about it

File: docs/api.rst [hunk 30]
```diff
@@ -2823,6 +2823,57 @@ Querying owned assets
...
+   .. sourcecode:: http
+
+      HTTP/1.1 200 OK
+      Content-Type: application/json
+
+      {
+          "result": {
+              "0x31F05553be0EBBf7774241603Cc7b28771F989B3": {
+                  "tokens": [
+                      "_ceth_0x6B175474E89094C44Da98b954EedeAC495271d0F", "_ceth_0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
```
Comment (LefterisJP) [inline docs/api.rst:2862]: nitpick:

```suggestion
                      "_ceth_0x6B175474E89094C44Da98b954EedeAC495271d0F",
                      "_ceth_0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
```

File: docs/api.rst [hunk 31]
```diff
@@ -2823,6 +2823,57 @@ Querying owned assets
...
+                  ],
+                  "last_update_timestamp": 1658764910,
```
Comment (LefterisJP) [inline docs/api.rst:2864]: invalid json

```suggestion
                  "last_update_timestamp": 1658764910
```

File: rotkehlchen/tests/integration/test_blockchain.py [hunk 32]
```diff
@@ -104,11 +106,20 @@ def mock_add_defi_balances_to_token_and_totals():
...
+            ethtokens.detect_tokens(
+                write_cursor=write_cursor,
+                only_cache=False,
+                accounts=[addr1, addr2],
+            )
+
     assert addr1 in blockchain.accounts.eth
 
     with etherscan_patch, ethtokens_max_chunks_patch, defi_balances_mock, add_defi_mock, beaconchain_patch:  # noqa: E501
+        blockchain.query_ethereum_balances(ignore_cache=True)
```
Comment (LefterisJP) [inline rotkehlchen/tests/integration/test_blockchain.py:120]: With this new approach of splitting token detection I am not so sure this makes much sense anymore though. So what are you doing here?

Detecting tokens at like 111? And then querying balances once without cache and then 5 times with cache? Why did you add the one time without cache in the beginning?

**Review by LefterisJP — APPROVED** [review]
I have a question and 2 nitpicks. Both can be addressed later.

You would also need to coordinate with frontend on how it should adjust to these changes.

Also would need you to do the change we discussed here: https://github.com/rotki/rotki/pull/4536#discussion_r928172015

which would be a really nice improvement.
